File: 1/AI1110/Random-Numbers/codes/3.1.c
// Generates 1000000 samples of a random variable V = -2ln(1-U) where U is a uniform random variable taking values in [0,1]

// Name: Ankit Saha
// Roll number: AI21BTECH11004

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define NUM 1000000

int main() {
	FILE *fp = fopen("v.dat", "w");
	if (!fp) {
		printf("Couldn't open file\n");
		return 1;
	}
	for (int i = 0; i < NUM; i++) {
		double u = (double)rand()/RAND_MAX;
		if (u >= 1) u = 0.999;
		double v = -2 * log(1 - u);
		fprintf(fp, "%lf\n", v);
	}
	fclose(fp);
	return 0;	
}

File: 1/AI1110/Random-Numbers/codes/3.1.py
# Loads the v.dat file and plots the empirical CDF of the random variable V

# Name: Ankit Saha
# Roll number: AI21BTECH11004

import numpy as np
import matplotlib.pyplot as plt

values = np.loadtxt('v.dat', dtype='double')
NUM = len(values)
X1 = np.linspace(-2, 5, 30)
X2 = np.linspace(-2, 5, 100)
emp_cdf = []
the_cdf = []

for x in X1:
    count = 0
    for value in values:
        if value <= x:
            count += 1
    emp_cdf.append(count/NUM)

for x in X2:
    if x < 0:
        the_cdf.append(0)
    else:
        the_cdf.append(1 - np.exp(-0.5 * x))

plt.plot(X2, the_cdf, label='Theoretical')
plt.plot(X1, emp_cdf, 'o', label='Empirical')
plt.grid()
plt.legend()
plt.savefig('../figs/3.1.png')
plt.show()
